src/Periphery/TokenWrapper.sol

```diff
@@ -1,59 +1,94 @@
 /// @author LI.FI (https://li.fi)
 /// @notice Provides functionality for wrapping and unwrapping tokens
-/// @custom:version 1.1.0
+/// @custom:version 1.2.0
 contract TokenWrapper is WithdrawablePeriphery {
-    uint256 private constant MAX_INT = 2 ** 256 - 1;
-    address public wrappedToken;
+    address public immutable WRAPPED_TOKEN;
+    address public immutable CONVERTER;
+    bool private immutable USE_CONVERTER;
+    uint256 private immutable SWAP_RATIO_MULTIPLIER;
 
     /// Errors ///
     error WithdrawFailure();
 
-    /// Constructor ///
+    /// @notice Creates a new TokenWrapper contract
+    /// @param _wrappedToken Address of the wrapped token (e.g., WETH, or token returned by converter)
+    /// @param _converter Address of converter contract, or address(0) if wrapping 1:1 without conversion
+    /// @param _owner Address that will own this contract and can withdraw stuck tokens
+    /// @dev If converter is provided, all wrap/unwrap operations go through it for decimal or other conversions
     // solhint-disable-next-line no-empty-blocks
     constructor(
         address _wrappedToken,
+        address _converter,
         address _owner
     ) WithdrawablePeriphery(_owner) {
-        wrappedToken = _wrappedToken;
-        IERC20(wrappedToken).approve(address(this), MAX_INT);
+        if (_wrappedToken == address(0)) revert InvalidConfig();
+        if (_owner == address(0)) revert InvalidConfig();
+
+        WRAPPED_TOKEN = _wrappedToken;
+        USE_CONVERTER = _converter != address(0);
+
+        if (USE_CONVERTER) {
+            if (!LibAsset.isContract(_converter)) revert InvalidContract();
+            CONVERTER = _converter;
+            // Approve converter once for all future withdrawals (gas optimization)
+            LibAsset.maxApproveERC20(
+                IERC20(_wrappedToken),
+                _converter,
+                type(uint256).max
+            );
+            // Calculate swap ratio based on decimal difference between native (18) and wrapped token
+            uint8 wrappedDecimals = IERC20Metadata(_wrappedToken).decimals();
+            SWAP_RATIO_MULTIPLIER = 10 ** wrappedDecimals;
+        } else {
+            CONVERTER = _wrappedToken;
+            SWAP_RATIO_MULTIPLIER = 1 ether; // 1:1 ratio for 18 decimals
+        }
     }
 
     /// External Methods ///
 
-    /// @notice Wraps the native token
+    /// @notice Wraps the native token and transfers wrapped tokens to caller
+    /// @dev If converter is set, uses it to convert native to wrapped tokens using precalculated ratio
+    /// @dev If no converter, wraps native 1:1 and transfers msg.value of wrapped tokens
     function deposit() external payable {
-        IWrapper(wrappedToken).deposit{ value: msg.value }();
-        IERC20(wrappedToken).transfer(msg.sender, msg.value);
+        IWrapper(CONVERTER).deposit{ value: msg.value }();
+        uint256 wrappedAmount = (msg.value * SWAP_RATIO_MULTIPLIER) / 1 ether;
```

Comment:
would recommend to use a speaking constant instead of `1 ether`, e.g.:
 BASE_DENOMINATOR
 
 same below